File: tools/python_bind/tests/test_explain_profile.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright 2020 Alibaba Group Holding Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""
End-to-end tests for PROFILE and EXPLAIN modes in both AP (OLAP) and TP (OLTP) modes.

This test suite covers:
- AP mode (local embedded): QueryProcessor::execute_internal
- TP mode (remote HTTP service): NeugDBSession::Eval
- Both modes sharing the same underlying Pipeline/OprTimer machinery

Tests verify:
1. PROFILE mode: Executes query and collects per-operator metrics
2. EXPLAIN mode: Builds operator tree without executing, returns 0 rows
3. Dict interface: get_profile_metrics() returns structured metrics
4. Text interface: get_profile_text() returns ASCII-formatted output
"""

import logging
import os
import shutil
import sys
import tempfile
import time

import pytest
from conftest import wait_for_server_ready

from neug.database import Database
from neug.session import Session

logger = logging.getLogger(__name__)


def load_sample_data(conn):
    """
    Create and populate a simple graph with person and software nodes.

    Args:
        conn: A connection object (either AP or TP mode) to execute queries.
    """
    conn.execute(
        "CREATE NODE TABLE person(id INT64, name STRING, age INT64, PRIMARY KEY(id));"
    )
    conn.execute(
        "CREATE NODE TABLE software(id INT64, name STRING, lang STRING, PRIMARY KEY(id));"
    )
    conn.execute("CREATE REL TABLE knows(FROM person TO person, weight DOUBLE);")
    conn.execute("CREATE REL TABLE created(FROM person TO software, weight DOUBLE);")

    # Insert person data
    conn.execute("CREATE (p:person {id: 1, name: 'marko', age: 29});")
    conn.execute("CREATE (p:person {id: 2, name: 'vadas', age: 27});")
    conn.execute("CREATE (p:person {id: 3, name: 'josh', age: 32});")
    conn.execute("CREATE (p:person {id: 4, name: 'peter', age: 35});")

    # Insert software data
    conn.execute("CREATE (s:software {id: 10, name: 'lop', lang: 'java'});")
    conn.execute("CREATE (s:software {id: 11, name: 'ripple', lang: 'java'});")

    # Insert relationships
    conn.execute(
        "MATCH (a:person), (b:person) WHERE a.id = 1 AND b.id = 2 "
        "CREATE (a)-[:knows {weight: 0.5}]->(b);"
    )
    conn.execute(
        "MATCH (a:person), (b:person) WHERE a.id = 1 AND b.id = 3 "
        "CREATE (a)-[:knows {weight: 1.0}]->(b);"
    )
    conn.execute(
        "MATCH (a:person), (b:person) WHERE a.id = 2 AND b.id = 3 "
        "CREATE (a)-[:knows {weight: 0.7}]->(b);"
    )
    conn.execute(
        "MATCH (a:person), (b:person) WHERE a.id = 3 AND b.id = 4 "
        "CREATE (a)-[:knows {weight: 0.4}]->(b);"
    )

    conn.execute(
        "MATCH (a:person), (s:software) WHERE a.id = 1 AND s.id = 10 "
        "CREATE (a)-[:created {weight: 0.4}]->(s);"
    )
    conn.execute(
        "MATCH (a:person), (s:software) WHERE a.id = 4 AND s.id = 10 "
        "CREATE (a)-[:created {weight: 0.6}]->(s);"
    )
    conn.execute(
        "MATCH (a:person), (s:software) WHERE a.id = 1 AND s.id = 11 "
        "CREATE (a)-[:created {weight: 1.0}]->(s);"
    )


@pytest.fixture(scope="module")
def ap_mode_db(tmp_path_factory):
    """
    Fixture for AP mode (embedded, local).

    Yields:
        conn: A connection object for executing queries in AP mode.
    """
    db_dir = str(tmp_path_factory.mktemp("test_explain_profile_ap"))
    db = Database(db_dir, "w")
    conn = db.connect()

    load_sample_data(conn)

    yield conn

    conn.close()
    db.close()


@pytest.fixture(scope="module")
def tp_mode_db(tmp_path_factory):
    """
    Fixture for TP mode (remote HTTP service).

    Yields:
        endpoint: The HTTP endpoint URI of the service.
    """
    db_dir = str(tmp_path_factory.mktemp("test_explain_profile_tp"))
    db = Database(db_dir, "w")
    conn = db.connect()
    load_sample_data(conn)
    conn.close()

    # Start the service
    endpoint = db.serve(10030, "localhost", False)
    time.sleep(1)

    yield endpoint

    db.stop_serving()
    db.close()


# ============================================================================
# AP Mode Tests
# ============================================================================


class TestAPModeProfile:
    """Tests for PROFILE mode in AP (embedded) mode."""

    def test_profile_single_table_scan(self, ap_mode_db):
        """Test PROFILE with single table scan."""
        conn = ap_mode_db
        query = "PROFILE MATCH (p:person) RETURN p.name, p.age"
        result = conn.execute(query)

        # Verify query results
        rows = list(result)
        assert len(rows) == 4, "Expected 4 person records"
        col_names = result.column_names()
        assert len(col_names) == 2
        assert "name" in col_names[0]
        assert "age" in col_names[1]

        # Verify profile result exists
        assert result.has_profile_result(), "PROFILE result should be present"

        # Verify metrics dict
        metrics = result.get_profile_metrics()
        assert "total_elapsed_ms" in metrics
        assert "total_output_rows" in metrics
        assert "operators" in metrics
        assert metrics["total_output_rows"] == 4
        assert len(metrics["operators"]) > 0

        # Verify text output
        text_output = result.get_profile_text()
        assert isinstance(text_output, str)
        assert len(text_output) > 0
        logger.info(f"Profile output:\n{text_output}")

    def test_profile_with_join(self, ap_mode_db):
        """Test PROFILE with join operation."""
        conn = ap_mode_db
        query = "PROFILE MATCH (p1:person), (p2:person) WHERE p1.id < p2.id RETURN p1.name, p2.name"
        result = conn.execute(query)

        # Verify query results
        rows = list(result)
        assert len(rows) == 6, "Expected 6 cross-product pairs (4 choose 2)"

        # Verify profile result
        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert metrics["total_output_rows"] == 6

        logger.info(
            f"Join profile: {metrics['total_elapsed_ms']:.3f} ms, {metrics['total_output_rows']} rows"
        )

    def test_profile_with_aggregation(self, ap_mode_db):
        """Test PROFILE with aggregation."""
        conn = ap_mode_db
        query = "PROFILE MATCH (p:person) RETURN COUNT(*) as person_count"
        result = conn.execute(query)

        # Verify query results
        rows = list(result)
        assert len(rows) == 1
        assert rows[0][0] == 4

        # Verify profile result
        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert metrics["total_output_rows"] == 1

        logger.info(f"Aggregation profile: {metrics['total_elapsed_ms']:.3f} ms")

    def test_profile_with_edge_traversal(self, ap_mode_db):
        """Test PROFILE with edge traversal."""
        conn = ap_mode_db
        query = "PROFILE MATCH (p:person)-[e:knows]->(q:person) RETURN p.name, q.name"
        result = conn.execute(query)

        # Verify query results
        rows = list(result)
        assert len(rows) == 4, "Expected 4 knows relationships"

        # Verify profile result
        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert metrics["total_output_rows"] == 4

        logger.info(f"Edge traversal profile: {metrics['total_elapsed_ms']:.3f} ms")

    def test_profile_metrics_structure(self, ap_mode_db):
        """Test the structure of profile metrics dict."""
        conn = ap_mode_db
        query = "PROFILE MATCH (p:person) RETURN p.id LIMIT 2"
        result = conn.execute(query)

        list(result)  # Consume result

        metrics = result.get_profile_metrics()

        # Check top-level keys
        assert "total_elapsed_ms" in metrics
        assert "total_output_rows" in metrics
        assert "operators" in metrics

        # Check operator structure
        assert len(metrics["operators"]) > 0
        for op in metrics["operators"]:
            assert "operator_id" in op
            assert "operator_name" in op
            assert "elapsed_ms" in op
            assert "output_rows" in op
            assert "parent_id" in op
            assert "child_ids" in op

            logger.info(
                f"Operator {op['operator_id']}: {op['operator_name']} "
                f"({op['elapsed_ms']:.3f} ms, {op['output_rows']} rows)"
            )


class TestAPModeExplain:
    """Tests for EXPLAIN mode in AP (embedded) mode."""

    def test_explain_basic(self, ap_mode_db):
        """Test EXPLAIN mode returns plan but no data rows."""
        conn = ap_mode_db
        query = "EXPLAIN MATCH (p:person)-[e:knows]->(q:person) RETURN p.name, q.name"
        result = conn.execute(query)

        # EXPLAIN should return 0 rows
        rows = list(result)
        assert len(rows) == 0, "EXPLAIN should return 0 data rows"

        # But should have profile result (the plan)
        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert metrics["total_output_rows"] == 0
        assert len(metrics["operators"]) > 0

        logger.info(f"Explain plan has {len(metrics['operators'])} operators")

    def test_explain_complex_query(self, ap_mode_db):
        """Test EXPLAIN with complex query."""
        conn = ap_mode_db
        query = (
            "EXPLAIN MATCH (p:person)-[e1:knows]->(q:person), "
            "(q)-[e2:created]->(s:software) "
            "RETURN p.name, s.name"
        )
        result = conn.execute(query)

        # EXPLAIN returns no rows
        rows = list(result)
        assert len(rows) == 0

        # But plan is visible
        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert len(metrics["operators"]) > 0

        text_output = result.get_profile_text()
        assert len(text_output) > 0
        logger.info(f"Complex plan:\n{text_output}")

    def test_explain_text_output_format(self, ap_mode_db):
        """Test EXPLAIN text output has proper formatting."""
        conn = ap_mode_db
        query = "EXPLAIN MATCH (p:person) RETURN p.id"
        result = conn.execute(query)

        list(result)  # Consume result

        text_output = result.get_profile_text()

        # Should be non-empty string
        assert isinstance(text_output, str)
        assert len(text_output) > 0

        # Should contain box-drawing characters or ASCII formatting
        assert any(char in text_output for char in ["─", "│", "┌", "└", "+", "|", "-"])

        logger.info(f"Text output:\n{text_output}")


class TestAPModeNormalExecution:
    """Tests for normal execution (no PROFILE/EXPLAIN) in AP mode."""

    def test_normal_execution_no_profile(self, ap_mode_db):
        """Test normal query execution without PROFILE/EXPLAIN."""
        conn = ap_mode_db
        query = "MATCH (p:person) RETURN p.name"
        result = conn.execute(query)

        rows = list(result)
        assert len(rows) == 4

        # Normal execution should not have profile result
        assert not result.has_profile_result()


# ============================================================================
# TP Mode Tests (HTTP Remote Service)
# ============================================================================


class TestTPModeProfile:
    """Tests for PROFILE mode in TP (remote service) mode."""

    def test_profile_single_table_scan(self, tp_mode_db):
        """Test PROFILE with single table scan in TP mode."""
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        query = "PROFILE MATCH (p:person) RETURN p.name, p.age"
        result = session.execute(query)

        # Verify query results
        rows = list(result)
        assert len(rows) == 4, "Expected 4 person records"
        col_names = result.column_names()
        assert len(col_names) == 2
        assert "name" in col_names[0]
        assert "age" in col_names[1]

        # Verify profile result exists
        assert result.has_profile_result(), "PROFILE result should be present"

        # Verify metrics dict
        metrics = result.get_profile_metrics()
        assert "total_elapsed_ms" in metrics
        assert "total_output_rows" in metrics
        assert "operators" in metrics
        assert metrics["total_output_rows"] == 4
        assert len(metrics["operators"]) > 0

        # Verify text output
        text_output = result.get_profile_text()
        assert isinstance(text_output, str)
        assert len(text_output) > 0
        logger.info(f"TP Profile output:\n{text_output}")

        session.close()

    def test_profile_with_join(self, tp_mode_db):
        """Test PROFILE with join operation in TP mode."""
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        query = "PROFILE MATCH (p1:person), (p2:person) WHERE p1.id < p2.id RETURN p1.name, p2.name"
        result = session.execute(query)

        rows = list(result)
        assert len(rows) == 6

        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert metrics["total_output_rows"] == 6

        logger.info(f"TP Join profile: {metrics['total_elapsed_ms']:.3f} ms")

        session.close()

    def test_profile_with_edge_traversal(self, tp_mode_db):
        """Test PROFILE with edge traversal in TP mode."""
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        query = "PROFILE MATCH (p:person)-[e:knows]->(q:person) RETURN p.name, q.name"
        result = session.execute(query)

        rows = list(result)
        assert len(rows) == 4

        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert metrics["total_output_rows"] == 4

        logger.info(f"TP Edge traversal profile: {metrics['total_elapsed_ms']:.3f} ms")

        session.close()


class TestTPModeExplain:
    """Tests for EXPLAIN mode in TP (remote service) mode."""

    def test_explain_basic(self, tp_mode_db):
        """Test EXPLAIN mode in TP returns plan but no data rows."""
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        query = "EXPLAIN MATCH (p:person)-[e:knows]->(q:person) RETURN p.name, q.name"
        result = session.execute(query)

        # EXPLAIN should return 0 rows
        rows = list(result)
        assert len(rows) == 0, "EXPLAIN should return 0 data rows"

        # But should have profile result (the plan)
        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert metrics["total_output_rows"] == 0
        assert len(metrics["operators"]) > 0

        logger.info(f"TP Explain plan has {len(metrics['operators'])} operators")

        session.close()

    def test_explain_complex_query(self, tp_mode_db):
        """Test EXPLAIN with complex query in TP mode."""
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        query = (
            "EXPLAIN MATCH (p:person)-[e1:knows]->(q:person), "
            "(q)-[e2:created]->(s:software) "
            "RETURN p.name, s.name"
        )
        result = session.execute(query)

        rows = list(result)
        assert len(rows) == 0

        assert result.has_profile_result()
        metrics = result.get_profile_metrics()
        assert len(metrics["operators"]) > 0

        text_output = result.get_profile_text()
        assert len(text_output) > 0
        logger.info(f"TP Complex plan:\n{text_output}")

        session.close()


class TestTPModeNormalExecution:
    """Tests for normal execution (no PROFILE/EXPLAIN) in TP mode."""

    def test_normal_execution_no_profile(self, tp_mode_db):
        """Test normal query execution without PROFILE/EXPLAIN in TP mode."""
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        query = "MATCH (p:person) RETURN p.name"
        result = session.execute(query)

        rows = list(result)
        assert len(rows) == 4

        # Normal execution should not have profile result
        assert not result.has_profile_result()

        session.close()


# ============================================================================
# Cross-Mode Consistency Tests
# ============================================================================


class TestAPTPConsistency:
    """Tests to verify AP and TP modes produce consistent results."""

    def test_profile_result_count_consistency(self, ap_mode_db, tp_mode_db):
        """Verify AP and TP produce same row count for PROFILE."""
        query = "PROFILE MATCH (p:person) RETURN p.id"

        # AP mode
        result_ap = ap_mode_db.execute(query)
        rows_ap = list(result_ap)
        metrics_ap = result_ap.get_profile_metrics()

        # TP mode
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        result_tp = session.execute(query)
        rows_tp = list(result_tp)
        metrics_tp = result_tp.get_profile_metrics()
        session.close()

        # Both should return 4 rows
        assert len(rows_ap) == len(rows_tp) == 4
        assert metrics_ap["total_output_rows"] == metrics_tp["total_output_rows"] == 4

        logger.info(
            f"AP: {metrics_ap['total_output_rows']} rows, "
            f"TP: {metrics_tp['total_output_rows']} rows - CONSISTENT"
        )

    def test_explain_plan_consistency(self, ap_mode_db, tp_mode_db):
        """Verify AP and TP produce same plan structure for EXPLAIN."""
        query = "EXPLAIN MATCH (p:person)-[e:knows]->(q:person) RETURN p.name"

        # AP mode
        result_ap = ap_mode_db.execute(query)
        list(result_ap)
        metrics_ap = result_ap.get_profile_metrics()

        # TP mode
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        result_tp = session.execute(query)
        list(result_tp)
        metrics_tp = result_tp.get_profile_metrics()
        session.close()

        # Both should have same number of operators in plan
        assert len(metrics_ap["operators"]) == len(metrics_tp["operators"])
        assert metrics_ap["total_output_rows"] == metrics_tp["total_output_rows"] == 0

        logger.info(
            f"AP plan: {len(metrics_ap['operators'])} operators, "
            f"TP plan: {len(metrics_tp['operators'])} operators - CONSISTENT"
        )

    def test_profile_operator_names_consistency(self, ap_mode_db, tp_mode_db):
        """Verify AP and TP produce same operator names."""
        query = "PROFILE MATCH (p:person) RETURN p.name"

        # AP mode
        result_ap = ap_mode_db.execute(query)
        list(result_ap)
        metrics_ap = result_ap.get_profile_metrics()
        ops_ap = [op["operator_name"] for op in metrics_ap["operators"]]

        # TP mode
        endpoint = tp_mode_db
        session = Session.open(endpoint)
        result_tp = session.execute(query)
        list(result_tp)
        metrics_tp = result_tp.get_profile_metrics()
        ops_tp = [op["operator_name"] for op in metrics_tp["operators"]]
        session.close()

        # Operator sequence should match
        assert ops_ap == ops_tp

        logger.info(f"Operators: {' -> '.join(ops_ap)} - CONSISTENT")

File: tests/unittest/test_explain_profile.cpp
/**
 * Copyright 2020 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>

#include <chrono>
#include <filesystem>
#include <memory>
#include <string>

#include "column_assertions.h"
#include "neug/generated/proto/response/response.pb.h"
#include "neug/main/connection.h"
#include "neug/main/neug_db.h"
#include "utils.h"

namespace neug {
namespace test {

/**
 * ExplainProfileTest: Tests for PROFILE and EXPLAIN execution modes
 *
 * This test suite verifies:
 * 1. Query execution returns ProfileResult in PROFILE mode
 * 2. Query execution returns execution tree in EXPLAIN mode (no data)
 * 3. ProfileResult structure is correctly populated
 */
class ExplainProfileTest : public ::testing::Test {
 protected:
  void SetUp() override {
    std::string dir_name =
        "neug_test_explain_profile_" +
        std::to_string(
            std::chrono::steady_clock::now().time_since_epoch().count());
    test_dir_ = std::filesystem::temp_directory_path() / dir_name;
    if (std::filesystem::exists(test_dir_)) {
      std::filesystem::remove_all(test_dir_);
    }
    std::filesystem::create_directories(test_dir_);
    db_path_ = (test_dir_ / "graph").string();

    // Load modern graph with a local DB instance, then close it.
    // Each test reopens a fresh DB — same pattern as test_checkpoint.cc.
    neug::NeugDB db;
    db.Open(db_path_, 4);
    auto conn = db.Connect();
    load_modern_graph(conn);
    conn->Close();
    db.Close();
  }

  void TearDown() override {
    if (std::filesystem::exists(test_dir_)) {
      std::filesystem::remove_all(test_dir_);
    }
  }

  std::string db_path_;
  std::filesystem::path test_dir_;
};

// ============================================================================
// Test Group 1: Regular Query (without PROFILE/EXPLAIN)
// ============================================================================

// Test 1.1: Regular query without PROFILE should not have ProfileResult
TEST_F(ExplainProfileTest, RegularQueryWithoutProfile) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();

  auto result = conn->Query("MATCH (n:person) RETURN n.name LIMIT 1");
  EXPECT_TRUE(result) << "Regular query should execute successfully";

  if (result) {
    const auto& response = result.value().response();

    // Verify regular query doesn't include ProfileResult
    if (response.has_profile_result()) {
      EXPECT_EQ(response.profile_result().operators_size(), 0)
          << "Regular query should not build operator tree";
    }

    // Verify response structure
    EXPECT_EQ(response.row_count(), 1) << "LIMIT 1 should return exactly 1 row";
    EXPECT_EQ(response.arrays_size(), 1)
        << "SELECT n.name should return 1 column";

    // Verify data content: first person is marko (id=1, name="marko")
    neug::test::AssertStringColumn(response, 0, {"marko"});
  }
  conn->Close();
  db.Close();
}

// Test 1.2: Regular query returns actual data
TEST_F(ExplainProfileTest, RegularQueryReturnsData) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();

  auto result = conn->Query("MATCH (n:person) RETURN n.name");
  EXPECT_TRUE(result);

  if (result) {
    const auto& response = result.value().response();

    // Verify response structure
    EXPECT_EQ(response.row_count(), 4) << "Should return all 4 person nodes";
    EXPECT_EQ(response.arrays_size(), 1)
        << "SELECT n.name should return 1 column";

    // Verify data content: person names in insertion order
    neug::test::AssertStringColumn(response, 0,
                                   {"marko", "vadas", "josh", "peter"});
  }
  conn->Close();
  db.Close();
}

// ============================================================================
// Test Group 2: PROFILE Mode - Verify ProfileResult is Generated
// ============================================================================

// Test 2.1: Regular query with PROFILE returns ProfileResult
TEST_F(ExplainProfileTest, ProfileModeSingleScan) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();

  auto result = conn->Query("PROFILE MATCH (n:person) RETURN n.name");
  EXPECT_TRUE(result) << "Query execution should succeed";

  if (result) {
    const auto& response = result.value().response();
    EXPECT_TRUE(response.has_profile_result())
        << "PROFILE mode should generate ProfileResult";
    if (response.has_profile_result()) {
      const auto& profile = response.profile_result();
      EXPECT_GT(profile.operators_size(), 0)
          << "ProfileResult should contain operator metrics";
      EXPECT_GE(profile.total_elapsed_ms(), 0)
          << "total_elapsed_ms should be non-negative";

      // Print profile_result_text() for inspection
      std::cout << "\n=== PROFILE OUTPUT ===\n"
                << result.value().profile_result_text()
                << "=== END PROFILE ===\n"
                << std::endl;
    }
  }
  conn->Close();
  db.Close();
}

// Test 2.2: Join query with PROFILE (cross product join)
TEST_F(ExplainProfileTest, ProfileModeWithJoin) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();

  // Real join: two independent MATCH clauses create cross product
  // MATCH (n:person), (m:person) produces 4*4=16 result rows
  auto result =
      conn->Query("PROFILE MATCH (n:person), (m:person) RETURN n.name, m.name");
  EXPECT_TRUE(result) << "Join query execution should succeed";

  if (result) {
    const auto& response = result.value().response();
    EXPECT_TRUE(response.has_profile_result())
        << "PROFILE mode with join should generate ProfileResult";
    if (response.has_profile_result()) {
      const auto& profile = response.profile_result();
      EXPECT_GT(profile.operators_size(), 0)
          << "ProfileResult should show multiple operators for join";
      EXPECT_GE(profile.total_output_rows(), 16)
          << "Join of 4 persons should produce at least 16 rows";

      // Print profile_result_text() for inspection
      std::cout << "\n=== PROFILE OUTPUT (Cross Product Join) ===\n"
                << result.value().profile_result_text()
                << "=== END PROFILE ===\n"
                << std::endl;
    }
  }
  conn->Close();
  db.Close();
}

// Test 2.3: Aggregation query with PROFILE
TEST_F(ExplainProfileTest, ProfileModeWithAggregation) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();

  auto result = conn->Query("PROFILE MATCH (n:person) RETURN COUNT(*) as cnt");
  EXPECT_TRUE(result) << "Aggregation query execution should succeed";

  if (result) {
    const auto& response = result.value().response();
    EXPECT_TRUE(response.has_profile_result())
        << "PROFILE mode with aggregation should generate ProfileResult";

    // Print profile_result_text() for inspection
    std::cout << "\n=== PROFILE OUTPUT ===\n"
              << result.value().profile_result_text() << "=== END PROFILE ===\n"
              << std::endl;
  }
  conn->Close();
  db.Close();
}

// ============================================================================
// Test Group 3: EXPLAIN Mode - Verify Execution Tree is Built
// ============================================================================

// Test 3.1: EXPLAIN mode returns execution tree structure
TEST_F(ExplainProfileTest, ExplainModeSingleScan) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();

  auto result = conn->Query("EXPLAIN MATCH (n:person) RETURN n.name");
  EXPECT_TRUE(result) << "EXPLAIN query execution should succeed";

  if (result) {
    const auto& response = result.value().response();
    EXPECT_TRUE(response.has_profile_result())
        << "EXPLAIN mode should generate ProfileResult with execution tree";
    EXPECT_EQ(response.row_count(), 0)
        << "EXPLAIN mode should not return data rows";
    if (response.has_profile_result()) {
      const auto& profile = response.profile_result();
      EXPECT_GT(profile.operators_size(), 0)
          << "EXPLAIN mode should show operator tree";
      EXPECT_EQ(profile.total_elapsed_ms(), 0)
          << "EXPLAIN mode should have 0 elapsed time (no execution)";
      EXPECT_EQ(profile.total_output_rows(), 0)
          << "EXPLAIN mode should have 0 output rows (no execution)";

      // Print profile_result_text() for inspection
      std::cout << "\n=== PROFILE OUTPUT ===\n"
                << result.value().profile_result_text()
                << "=== END PROFILE ===\n"
                << std::endl;
    }
  }
  conn->Close();
  db.Close();
}

// Test 3.2: EXPLAIN with join (cross product join)
TEST_F(ExplainProfileTest, ExplainModeWithJoin) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();
  ASSERT_NE(conn, nullptr);

  // Real join: EXPLAIN shows query plan without executing it
  auto result =
      conn->Query("EXPLAIN MATCH (n:person), (m:person) RETURN n.name, m.name");
  EXPECT_TRUE(result) << "EXPLAIN join query should succeed";

  if (result) {
    const auto& response = result.value().response();
    EXPECT_TRUE(response.has_profile_result())
        << "EXPLAIN should generate execution tree";
    EXPECT_EQ(response.row_count(), 0)
        << "EXPLAIN should not execute or return rows";
    if (response.has_profile_result()) {
      const auto& profile = response.profile_result();
      EXPECT_GT(profile.operators_size(), 0)
          << "EXPLAIN should show join operator tree";
      EXPECT_EQ(profile.total_elapsed_ms(), 0)
          << "EXPLAIN should not execute, so elapsed time is 0";
      EXPECT_EQ(profile.total_output_rows(), 0)
          << "EXPLAIN should not execute, so no output rows";

      // Print profile_result_text() for inspection
      std::cout << "\n=== PROFILE OUTPUT ===\n"
                << result.value().profile_result_text()
                << "=== END PROFILE ===\n"
                << std::endl;
    }
  }
  conn->Close();
  db.Close();
}

// ============================================================================
// Test Group 4: ProfileResult Structure Validation
// ============================================================================

// Test 4.1: ProfileResult contains required fields
TEST_F(ExplainProfileTest, ProfileResultFieldsPresent) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();

  auto result = conn->Query("PROFILE MATCH (n:person) RETURN n.age");
  EXPECT_TRUE(result);

  if (result) {
    const auto& response = result.value().response();
    if (response.has_profile_result()) {
      const auto& profile = response.profile_result();
      EXPECT_GT(profile.operators_size(), 0);
      for (int i = 0; i < profile.operators_size(); ++i) {
        const auto& op = profile.operators(i);
        EXPECT_FALSE(op.operator_name().empty())
            << "Operator name should not be empty";
        (void) op.operator_id();
        (void) op.parent_id();
        (void) op.elapsed_ms();
        (void) op.output_rows();
      }

      // Print profile_result_text() for inspection
      std::cout << "\n=== PROFILE OUTPUT ===\n"
                << result.value().profile_result_text()
                << "=== END PROFILE ===\n"
                << std::endl;
    }
  }
  conn->Close();
  db.Close();
}

// Test 4.2: Verify ProfileResult metrics accumulate correctly
TEST_F(ExplainProfileTest, ProfileResultMetricsAccumulate) {
  neug::NeugDB db;
  db.Open(db_path_, 4);
  auto conn = db.Connect();
  ASSERT_NE(conn, nullptr);

  auto result = conn->Query("PROFILE MATCH (n:person) RETURN n.name, n.age");
  EXPECT_TRUE(result);

  if (result) {
    const auto& response = result.value().response();
    if (response.has_profile_result()) {
      const auto& profile = response.profile_result();
      EXPECT_GE(profile.total_output_rows(),
                static_cast<uint64_t>(response.row_count()))
          << "ProfileResult total_output_rows should be >= returned rows";
      EXPECT_GT(profile.total_elapsed_ms(), 0)
          << "PROFILE execution should have non-zero elapsed time";

      // Print profile_result_text() for inspection
      std::cout << "\n=== PROFILE OUTPUT ===\n"
                << result.value().profile_result_text()
                << "=== END PROFILE ===\n"
                << std::endl;
    }
  }
  conn->Close();
  db.Close();
}

}  // namespace test
}  // namespace neug
